### PR TITLE
enhance: Add delete buffer related quota logic

### PR DESCRIPTION
**Pull request by congqixia**
See also #35303

### CODE REVIEW

File: configs/milvus.yaml [hunk 1]
```diff
@@ -979,6 +979,14 @@ quotaAndLimits:
...
       lowWaterLevel: 32768 # l0 segment row count quota, low water level
       highWaterLevel: 65536 # l0 segment row count quota, low water level
+    deleteBufferRowCountProtection:
+      enabled: false # switch to enable delete buffer row count quota
+      lowWaterLevel: 32768 # delete buffer row count quota, low water level
+      highWaterLevel: 65536 # delete buffer row count quota, high water level
+    deleteBufferSizeProtection:
+      enabled: false # switch to enable delete buffer size quota
+      lowWaterLevel: 32768 # delete buffer size quota, low water level
+      highWaterLevel: 65536 # delete buffer size quota, high water level
```
Comment (tedxu) [inline configs/milvus.yaml:989]: The default values may be too small, I suppose it won't be a problem if the deletions can be easily consumed in less than 1 minute or so.

And, please note the unit of the delete buffer size in the comments.

File: pkg/util/paramtable/quota_param.go [hunk 2]
```diff
@@ -1915,6 +1921,60 @@ but the rate will not be lower than minRateRatio * dmlRate.`,
...
+		DefaultValue: "false",
+		Doc:          "switch to enable delete buffer size quota",
+		Export:       true,
+	}
+	p.DeleteBufferSizeProtectionEnabled.Init(base.mgr)
+
+	p.DeleteBufferSizeLowWaterLevel = ParamItem{
+		Key:          "quotaAndLimits.limitWriting.deleteBufferSizeProtection.lowWaterLevel",
+		Version:      "2.4.11",
+		DefaultValue: "134217728", // 128MB
```
Comment (tedxu) [inline pkg/util/paramtable/quota_param.go:1963]: The default value is different from milvus.yaml, to correct that, you can delete the entries in milvus.yaml and re-generate it.

**Review by tedxu — COMMENTED** [review]
The patch looks good to me, except some minor issues.

Question: will this patch deprecate the same protection feature based on L0 row number?

File: pkg/util/paramtable/quota_param.go (hunk 2 above)
Comment (congqixia) [reply pkg/util/paramtable/quota_param.go:1963]: done

**Comment by congqixia** [comment]
> Question: will this patch deprecate the same protection feature based on L0 row number?

nope, this PR only add quota rules based on delete buffer number/size on querynodes, which will not affect the previously added l0 segment statistics one

File: configs/milvus.yaml [hunk 3]
```diff
@@ -979,6 +979,14 @@ quotaAndLimits:
       enabled: false # switch to enable l0 segment row count quota
       lowWaterLevel: 32768 # l0 segment row count quota, low water level
       highWaterLevel: 65536 # l0 segment row count quota, low water level
+    deleteBufferRowCountProtection:
+      enabled: false # switch to enable delete buffer row count quota
+      lowWaterLevel: 32768 # delete buffer row count quota, low water level
+      highWaterLevel: 65536 # delete buffer row count quota, high water level
```
Comment (XuanYang-cn) [inline configs/milvus.yaml:985]: This can be merged ahead, but the waterlevel is not a sensible one, need to adjust it before put in prod

Comment (congqixia) [reply configs/milvus.yaml:985]: agree, this value shall be updated after we get some experience data from the tests

**Pull request merged**